dubbo-remoting/dubbo-remoting-http12/src/main/java/org/apache/dubbo/remoting/http12/h2/Http2ServerChannelObserver.java

```diff
@@ -74,12 +76,33 @@
 
     @Override
     public void cancel(Throwable throwable) {
+        if (throwable instanceof CancelStreamException) {
+            if (((CancelStreamException) throwable).isCancelByRemote()) {
+                closed = true;
```

Comment:
**Performance Issue**: The condition checking and setting of the 'closed' flag based on the type of exception can lead to race conditions in a concurrent environment, potentially causing some streams to not be closed properly. <br> **Fix**: Ensure thread safety by synchronizing access to the 'closed' flag or by using atomic variables. Additionally, consider using more specific exception handling to manage stream closures more effectively. <br> **Code Suggestion**: 
 ```
 + private final AtomicBoolean closed = new AtomicBoolean(false);

 + if (throwable instanceof CancelStreamException && ((CancelStreamException) throwable).isCancelByRemote()) {
 +    closed.set(true);
 + }
 ```